horaemeta/server/coordinator/inspector/node_inspector.go

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package inspector
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/apache/incubator-horaedb-meta/pkg/log"
+	"github.com/apache/incubator-horaedb-meta/server/cluster/metadata"
+	"github.com/apache/incubator-horaedb-meta/server/storage"
+	"go.uber.org/zap"
+)
+
+const inspectInterval = time.Second * 5
+
+// NodeInspector will inspect node status and remove expired data.
+type NodeInspector struct {
+	logger          *zap.Logger
+	clusterMetadata *metadata.ClusterMetadata
+
+	// This lock is used to protect the following field.
+	lock      sync.RWMutex
+	isRunning atomic.Bool
+}
+
+func NewNodeInspector(logger *zap.Logger, clusterMetadata *metadata.ClusterMetadata) *NodeInspector {
+	return &NodeInspector{
+		logger:          logger,
+		clusterMetadata: clusterMetadata,
+		lock:            sync.RWMutex{},
+		isRunning:       atomic.Bool{},
+	}
+}
+
+func (i *NodeInspector) Start(ctx context.Context) error {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+
+	if i.isRunning.Load() {
+		return nil
+	}
```

Comment:
If you want to ensure the forloop to be started once, `sync.Once` may be a better choice. BTW, current way can't ensure that the following forloop is started only once.

Reply:
`isRunning` is used to cancel the loop rather than started only once, and the lock is needed for ensure only one caller can enter this loop at the same time, it is ensure that the following forloop is started only once.

Reply:
> the lock is needed for ensure only one caller can enter this loop at the same time, it is ensure that the following forloop is started only once.

I guess the lock can't make it because the goroutine may not be started when the lock is released.

Reply:
I've fixed it in the newest commints.